clarified Nursery.start vs Nursery.start_soon

rotu opened the pull request:
#970

oremanj commented on docs/source/history.rst:
```diff
@@ -5,6 +5,42 @@ Release history
 
 .. towncrier release notes start
 
+Trio 0.11.0+dev (2019-03-07)
```
The towncrier generated release notes shouldn't get checked in until we actually tag the release; the only change you would make is adding the new newsfragment file.

rotu replied:
Reverted

oremanj commented on docs/source/reference-core.rst:
```diff
@@ -834,13 +834,14 @@ Nursery objects provide the following interface:
-      If you want to wait for the task to initialize itself before 
-      continuing, see :meth:`start()`.
+      If you want to ensure that the child task does something before
+      before the parent moves on to the next line, or that multiple
```
duplicate "before"

oremanj commented on docs/source/reference-core.rst:
```diff
@@ -834,13 +834,14 @@ Nursery objects provide the following interface:
       then you don't need a nursery; just use ``await async_fn(*args)``.
-      If you want to wait for the task to initialize itself before 
-      continuing, see :meth:`start()`.
+      If you want to ensure that the child task does something before
```
What do you find to be clear from "ensure that the child task does something before the parent moves onto the next line" that isn't clear from "wait for the task to initialize itself before continuing"? I don't really see a substantive difference, and tend to prefer the shorter/punchier version, but I might be missing something. (I think being clear about "the *child* task" and maybe "before continuing *execution of the parent*" are worthwhile changes, but I'm sketchy on the more complex grammatical structure.)

rotu replied:
Initialize sounds technical, and implies to me that there's some sort of __init__ function being called, or that there's something other than a function being called. Rewrote to the punchier "If you want to run some part of the child task before the parent task proceeds, see :meth:`start()`."

oremanj commented on docs/source/reference-core.rst:
```diff
@@ -834,13 +834,14 @@ Nursery objects provide the following interface:
-      actually get a chance to do anything until some later point when
-      you execute a checkpoint and the scheduler decides to run it.
-      If you want to run a function and immediately wait for its result,
+      immediately, *before* it has a chance to run. At the next
```
I think this is a more stringent guarantee than we actually want to make -- #32 has some discussion of scheduling models that might want us to not switch tasks at every checkpoint.

rotu replied:
Rewrote to not guarantee that tasks will launch at the next checkpoint.

oremanj commented on newsfragments/970.doc.rst:
```diff
@@ -0,0 +1 @@
+Clarified documentation around Nursery.start_soon and Nursery.start, that start_soon does not guarantee task ordering, and when you should use each.
```
@njsmith - do we usually add newsfragments for documentation updates?

rotu replied:
@oremanj @njsmith Any opinions? I don't care whether we leave it in or take it out, but I'd like this pull request resolved

oremanj commented on docs/source/reference-core.rst:
```diff
@@ -878,9 +881,9 @@ Nursery objects provide the following interface:
 
          async def async_fn(arg1, arg2, *, task_status=trio.TASK_STATUS_IGNORED):
-             ...
+             ... # some work before the parent may proceed
```
A task with `task_status=trio.TASK_STATUS_IGNORED` can also be started with `start_soon()`, in which case the parts before and after `task_status.started()` behave the same as each other. Is there a different way to word this comment that might be clearer about that?

rotu replied:
reworded to "some work the parent might wait for"

rotu commented:
Thank you, @oremanj for your thorough and patient feedback. I'm not used to contributing to open source, and I appreciate your help greatly. Aside from the newsfragment note (which I'd be happy to go either way on) I think I addressed all your concerns.

oremanj commented on docs/source/reference-core.rst:
```diff
@@ -867,9 +867,9 @@ Nursery objects provide the following interface:
-      finished initializing itself, and optionally returns some
-      information from it.
+      Like :meth:`start_soon`, but blocks until the new child task
+      signals that it may unblock, optionally returning additional
```
"signals" is a slightly overloaded word to use here due to Unix signals. Maybe "until the new child task indicates that it's finished starting up"? And I think the original "and optionally returns some information" is better than "optionally returning additional information" -- the more active phrasing tends to communicate more clearly IMO.

rotu replied:
changing to "indicates"

oremanj commented on docs/source/reference-core.rst:
```diff
@@ -834,13 +834,13 @@ Nursery objects provide the following interface:
+      later), all tasks requested with :meth:`start_soon` will begin,
+      in no particular order. If you want to run a function to
+      completion, then you don't need a nursery; just use
+      ``await async_fn(*args)``. If you want to run some part of the
```
maybe "If you want to wait for the child task to finish starting up before the parent is allowed to proceed"

rotu replied:
No, because "finish starting up" is not a meaningful concept until you've already bought into the abstraction of Nursery.start(). A task scheduled with start_soon might never call a task_status.started() function - but any programmer would say that task starts and ends and infers that it "finishes starting" somewhere in between. There needs to be some concept of the child signaling to the parent, even if we don't use the word "signal".

oremanj commented on docs/source/reference-core.rst:
```diff
@@ -834,13 +834,13 @@ Nursery objects provide the following interface:
+      time, (which might be closing of the nursery context, but no
+      later), all tasks requested with :meth:`start_soon` will begin,
+      in no particular order. If you want to run a function to
+      completion, then you don't need a nursery; just use
```
The function always runs "to completion" (as opposed to being abandoned in the middle somewhere) regardless of whether you use `await` or `nursery.start_soon`. What are you trying to communicate here that you didn't feel was adequately communicated with "and immediately wait for its result"?

rotu replied:
The start function runs a function and immediately waits for a result too. It just so happens that the result might be returned by an intermediate result from a .started call in the middle.. I wanted to convey the difference. How about "If you want to run a function and block until it finishes" ?

oremanj commented on docs/source/reference-core.rst:
```diff
@@ -834,13 +834,13 @@ Nursery objects provide the following interface:
-      continuing, see :meth:`start()`.
+      immediately, *before* it has a chance to run. At some future
+      time, (which might be closing of the nursery context, but no
+      later), all tasks requested with :meth:`start_soon` will begin,
```
This still could be read to imply that all of the tasks will be started at once, which isn't a guarantee we want to make. What if anything do you think is unclear about taking the previous wording ("won't actually get a chance to do anything ... the scheduler decides to run it") and adding a sentence like "Tasks might start running in a different order than the calls to `start_soon()`"?

rotu replied:
You're right - it does imply a single time. Rephrased to "All tasks requested with :meth:`start_soon` will begin, at no particular time, in no particular order (but no later than the closing of the nursery context)."
I wanted to avoid mention of the scheduler here. The important thing is that ordering and timing get arbitrary when you call start_soon - I don't think the reader will get that unless they have a good grasp of what the scheduler does and does not promise.

rotu commented:
> sorry for being so nitpicky about this, but it's a pretty core aspect of the documentation

No problem! This documentation cost me personally hours of frustration and confusion. I'm more than happy to hammer it to death.

rotu commented:
What's left to fix for this PR?

pquentin commented:
@rotu I'm sorry that you had a bad experience here despite a very promising start. I personally don't feel qualified enough to look at this, but hopefully @oremanj will be able to look at this again at some point.
